### ISSUE TITLE
Test ConsoleTest.testConsoleHasOutput() is unstable

### ISSUE THREAD

**Issue by HannesWell** [eclipse-m2e/m2e-core#341]
The test case `ConsoleTest.testConsoleHasOutput()` seems to be unstable respectively to have a race-condition or comparable. Sometimes the test passes and sometimes not, even without changes. See for example https://ci.eclipse.org/m2e/job/m2e/job/master/89/ and https://ci.eclipse.org/m2e/job/m2e/job/master/90/ where the exact same commit was build. In Job 89 the test failed and in Job 90 it passed.

**Comment by laeubi** [comment]
One remark for the test: I think the very first thing to change here is, that instead of

```
System.out.println("Console Text: " + ((TextConsole)mavenConsole).getDocument().get());
assertTrue("missing output in console", ((TextConsole)mavenConsole).getDocument().get().contains("BUILD SUCCESS"));
```
we should replace with:

1. extract the `((TextConsole)mavenConsole).getDocument().get()` into a local variable so there is no chance in having different output in the debug and the assert
2. Instead of query the console once, we could simply have a loop that fetches the text, check if it contains the desired string and sleep for a few milliseconds and then recheck. If this fails for more than 10 times we could fail the test.
3. It woudl be good to make the failing message more expressive e.g. `fail("Expecting BUILD SUCCESS in console output but got "+text)`;

**Comment by laeubi** [comment]
Another aproach would be to add a [IPropertyChangeListener](https://javadoc.scijava.org/Eclipse/org/eclipse/ui/console/AbstractConsole.html#addPropertyChangeListener(org.eclipse.jface.util.IPropertyChangeListener)) and see if the text-console fires an event on text change, we can then use a countdonwn latch to signal the expected string was found + a timed wait on the latch.

**Comment by HannesWell** [comment]
I fully agree with your remarks. I would even suggest to use Hamcrest matchers.
Using a IPropertyChangeListener seems to be a more stable approach to me. In case a `IConsoleListener` is added to the consoleManager, the sleep-loop could be fully omitted.

I can take care of this, unless you want to prepare a PR?

**Comment by laeubi** [comment]
Just go a haed :-) I'm just not sure if the console emits an event on text change but we can try it out..

**Comment by laeubi** [comment]
Duplicate of #431

**Comment by vogella** [comment]
@HannesWell can you send the PR to stabilize the test?

**Comment by HannesWell** [comment]
> @HannesWell can you send the PR to stabilize the test?

It is not yet ready, but I will see what I can achieve this evening and submit it ASAP.

**Comment by vogella** [comment]
Thanks Hannes

Hannes Wellmann ***@***.***> schrieb am So., 12. Dez. 2021,
00:48:
